[4.1] Ravendb 10538

aviv86 opened the pull request.

aviv86 committed:
RavenDB-10538 : facets should use parameters (client side)

aviv86 committed:
RavenDB-10538 : facets should use parameters (server side)

aviv86 committed:
RavenDB-10538 : fix old facet tests, add new test cases

ayende commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -269,17 +276,25 @@ private static string GetStringRepresentation(string fieldName, ExpressionType l
+            var type = value.GetType().FullName;
+
+            if (parameters != null && 
+                type.In(new[] { "System.DateTime", "System.DateTimeOffset", "System.String" }))
```
Avoid doing things this way, instead, use an `is DateTime`, `is DateTimeOffset`, etc.

ayende replied:
Why is this needed, anyway?

ayende commented on src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs:
```diff
@@ -299,7 +318,38 @@ private static string ExtractFieldName(BinaryExpression be, FacetQuery query)
+                //and are not a part of the queryParameters
+
+                facetParameters.TryGetValue(parameter, out o);
+                var typeName = o?.GetType().FullName;
```
why are you passing the types in this manner?

aviv86 committed:
RavenDB-10538:
range facet: parse the Range-Expressions at the last possible point (where FacetToken is created),
in order to avoid rewriting the facet-parameters to query parameters and rewriting the ranges.

aviv86 committed:
RavenDB-10538 : fix tests
